COMPAT: Updated method call for _Openpyxl22Write.write_cells() for Openpyxl =>2.4.0 #14519

ditenberg opened the pull request:
- [X] closes #14519
- [X] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

Updated _Openpyxl22Writer.write_cells() to use new coordinate call method wks['coordinate']
instead of wks.cell() in accordance with https://openpyxl.readthedocs.io/en/default/tutorial.html#accessing-one-cell
in the newest release of Openpyxl (2.4.0 and later).

I did not know how to write a true test for this fix since it involves Excel file I/O and pandas does not currently supporting reading in .xlsx as far as I know. The best I could do is use it with openpyxl 2.4.0 install on my machine and passing a small dataframe into an .xlsx. Please let me know if you have any suggestions on how to better test this.

ditenberg committed:
COMPAT: openpyxl #14519

Updated _Openpyxl22Writer.write_cells() to use new coordinate call method wks['coordinate']
instead of wks.cell() in accordance with https://openpyxl.readthedocs.io/en/default/tutorial.html#accessing-one-cell
in the newest release of Openpyxl (2.4.0 and later).

ditenberg commented:
@jorisvandenbossche if you could point me to where you think the UserWarning might be stemming from in pandas codebase then I'd be more than happy to take a look at it. I just went the most obvious edit that I could find and only that used this wks.cell(coordinate) method call. If you know of any other sources for the issue please let me know.

TomAugspurger commented:
@ditenberg for finding the UserWarning, you could try sticking

```
import warnings
warnings.simplefilter('error', UserWarning)
```

at the top of the excel file tests. That would turn any UserWarnings into exceptions. Glancing through the test file, it looks like we use `.cell` in a few places: https://github.com/pandas-dev/pandas/blob/7f5a45c1b388b3f7f309f82bfa0733b7b9980c3a/pandas/io/tests/test_excel.py#L1970

ditenberg committed:
Update _Openpyxl20Writer(s)

Update _Openpyxl20Writer and fixed _Openpyxl22Writer write_cells() methods. Made 20 index cells using numeric row/columns. Made 22 the same way as well (original state in current master)

ditenberg committed:
Add exceptions to Openpyxl tests

Updated syntax to latest release openpyxl for retrieving cell values

ditenberg commented:
@TomAugspurger I put your suggested code into the pandas/io/tests/test_excel.py file and changed the cell value retrieval syntax in your mentioned location and a few others throughout the file. After running `nosetests pandas/io/tests/test_excel.py` I got back a clean report with no errors or UserWarnings.

ditenberg commented:
@TomAugspurger There are deeper issues with the _OpenpyxlXXWriter since they are deemed incompatible in the test_excel.py comments to be skipped during tests. Regardless, with commits the UserWarning no longer appears after test.

jreback commented on pandas/io/tests/test_excel.py:
```diff
@@ -2074,13 +2077,14 @@ def test_write_cells_merge_styled(self):
         ]
 
         with ensure_clean('.xlsx') as path:
+            warnings.simplefilter('error', UserWarning)
```
use assert_produces_warning

jreback commented on pandas/io/excel.py:
```diff
@@ -902,7 +902,7 @@ def write_cells(self, cells, sheet_name=None, startrow=0, startcol=0):
 
         for cell in cells:
             colletter = get_column_letter(startcol + cell.col + 1)
```
all of these changes need to be back-compat. so need to switch code based on the version.

jreback commented:
@ditenberg can you rebase / update.

this needs to be back compat (easy enough, you simply have a version check on the import and branch based on that).

jreback commented:
@ditenberg can you update this?

jreback commented:
closing, but would like you to update if possible.
